myia/api.py

```diff
@@ -895,6 +896,9 @@ def wrapped(*args):
             optlib.getitem_env_add,
             optlib.simplify_array_map,
         ],
+        unfuse=[
+            optlib.unfuse_composite,
+        ],
```

Comment:
I think I would put this as a separate opt step right before validate, because this transform is more a hindrance than an asset to the rest of the optimizer.

Reply:
I'm not so sure about that.  It is easier to reason about separate operations than a composite graph.

Reply:
The gradient doesn't generate a very nice graph as an argument to `array_map`, it generates something with calls to `J`, `make_tuple`, nested graphs, and so on, and then relies on the optimizer to clean it all up. It seems it gets cleaned up properly prior to `unfuse`, but I wouldn't rely on this.

Reply:
Then perhaps we should unfuse before the gradient happens.

That way the gradient should be relatively simple.

Reply:
Yeah, I noticed that issue too. `unfuse_composite` + `getitem_tuple` should be sufficient.